pkg/chubaofs/nodeserver.go

```diff
@@ -234,3 +243,148 @@ func nodeGetVolumeStats(_ context.Context, volumePath string) (*csi.NodeGetVolum
 		},
 	}, nil
 }
+
+// getAttachedPVOnNode finds all persistent volume objects attached in the node and controlled by me.
+func (ns *nodeServer) getAttachedPVOnNode(nodeName string) ([]*v1.PersistentVolume, error) {
+	vaList, err := ns.Driver.ClientSet.StorageV1().VolumeAttachments().List(metav1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("unable to list VolumeAttachments: %v", err)
+	}
+
+	nodePVNames := make(map[string]struct{})
+	for _, va := range vaList.Items {
+		if va.Spec.NodeName != nodeName {
+			continue
+		}
+		if va.Spec.Attacher != DriverName {
+			continue
+		}
+		if !va.Status.Attached {
+			continue
+		}
+		if va.Spec.Source.PersistentVolumeName == nil {
+			continue
+		}
+
+		nodePVNames[*va.Spec.Source.PersistentVolumeName] = struct{}{}
+	}
+
+	pvList, err := ns.Driver.ClientSet.CoreV1().PersistentVolumes().List(metav1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("unable to list PersistentVolumes: %v", err)
+	}
+
+	nodePVs := make([]*v1.PersistentVolume, 0, len(nodePVNames))
+	for i := range pvList.Items {
+		_, exist := nodePVNames[pvList.Items[i].Name]
+		if exist {
+			nodePVs = append(nodePVs, &pvList.Items[i])
+		}
+	}
+
+	return nodePVs, nil
+}
+
+type persistentVolumeWithPods struct {
+	*v1.PersistentVolume
+	pods []*v1.Pod
+}
+
+// getAttachedPVWithPodsOnNode finds all persistent volume objects as well as the related pods in the node.
+func (ns *nodeServer) getAttachedPVWithPodsOnNode(nodeName string) ([]*persistentVolumeWithPods, error) {
+	pvs, err := ns.getAttachedPVOnNode(nodeName)
+	if err != nil {
+		return nil, fmt.Errorf("getAttachedPVOnNode faied: %v", err)
+	}
+
+	claimedPVWithPods := make(map[string]*persistentVolumeWithPods, len(pvs))
+	for _, pv := range pvs {
+		if pv.Spec.ClaimRef == nil {
+			continue
+		}
+
+		pvcKey := fmt.Sprintf("%s/%s", pv.Spec.ClaimRef.Namespace, pv.Spec.ClaimRef.Name)
+		claimedPVWithPods[pvcKey] = &persistentVolumeWithPods{
+			PersistentVolume: pv,
+		}
+	}
+
+	allPodsOnNode, err := ns.Driver.ClientSet.CoreV1().Pods("").List(metav1.ListOptions{
+		FieldSelector: "spec.nodeName=" + nodeName,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("list pods failed: %v", err)
+	}
+
+	for i := range allPodsOnNode.Items {
+		pod := allPodsOnNode.Items[i]
+
+		for _, volume := range pod.Spec.Volumes {
+			if volume.PersistentVolumeClaim == nil {
+				continue
+			}
+			pvcKey := fmt.Sprintf("%s/%s", pod.Namespace, volume.PersistentVolumeClaim.ClaimName)
+			pvWithPods, ok := claimedPVWithPods[pvcKey]
+			if !ok {
+				continue
+			}
+
+			pvWithPods.pods = append(pvWithPods.pods, &pod)
```

Comment:
prefer to add 'break', no need check other volumes.

Reply:
done.